File: frontend/relief/src/services/reliefEngine.ts
import { 
  InventoryItem, 
  ReliefRequest, 
  MatchResult, 
  RegionalCoverageScore,
  VulnerableDemographics
} from '../types/relief';

const EARTH_RADIUS_KM = 6371;

export function haversineDistance(lat1: number, lon1: number, lat2: number, lon2: number): number {
  const dLat = (lat2 - lat1) * Math.PI / 180;
  const dLon = (lon2 - lon1) * Math.PI / 180;
  const a = 
    Math.sin(dLat / 2) * Math.sin(dLat / 2) +
    Math.cos(lat1 * Math.PI / 180) * Math.cos(lat2 * Math.PI / 180) * 
    Math.sin(dLon / 2) * Math.sin(dLon / 2);
  const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
  return Math.round(EARTH_RADIUS_KM * c * 10) / 10;
}

export function computePriorityScore(
  population: number,
  urgency: 'LOW' | 'MODERATE' | 'HIGH' | 'CRITICAL',
  vulnerable: VulnerableDemographics,
  roadBlocked: boolean,
  shortageRatio: number = 0.5
): { score: number; priorityClass: 'LOW' | 'MODERATE' | 'HIGH' | 'CRITICAL' } {
  // Population factor: max 25 points
  const popScore = Math.min(25, (Math.log10(Math.max(population, 1)) / 4.0) * 25.0);

  // Urgency weight: max 25 points
  const urgencyWeight = {
    'CRITICAL': 25,
    'HIGH': 18,
    'MODERATE': 10,
    'LOW': 4
  }[urgency];

  // Shortage score: max 20 points
  const shortageScore = Math.min(20, shortageRatio * 20.0);

  // Vulnerability factor: max 20 points
  const children = vulnerable.children || 0;
  const elderly = vulnerable.elderly || 0;
  const pregnant = vulnerable.pregnant || 0;
  const injured = vulnerable.injured || 0;
  const totalVuln = children + elderly + pregnant + injured;

  let vulnScore = 0;
  if (population > 0 && totalVuln > 0) {
    const weightedVuln = (children * 1.5 + elderly * 1.5 + pregnant * 2.5 + injured * 3.0);
    vulnScore = Math.min(20, (weightedVuln / population) * 40.0);
  }

  // Accessibility: 10 if cut off, 2 if accessible
  const accessibilityScore = roadBlocked ? 10 : 2;

  const total = Math.min(100, Math.max(0, Math.round((popScore + urgencyWeight + shortageScore + vulnScore + accessibilityScore) * 10) / 10));

  let priorityClass: 'LOW' | 'MODERATE' | 'HIGH' | 'CRITICAL' = 'LOW';
  if (total >= 80) priorityClass = 'CRITICAL';
  else if (total >= 60) priorityClass = 'HIGH';
  else if (total >= 35) priorityClass = 'MODERATE';

  return { score: total, priorityClass };
}

export function matchNeedToRelief(
  request: ReliefRequest,
  inventory: InventoryItem[]
): MatchResult {
  // Find all depots that hold the requested item or same category
  const candidates = inventory
    .filter(item => item.item_name.toLowerCase().includes(request.required_item.toLowerCase()) || 
                    item.category === request.category)
    .filter(item => item.quantity > 0)
    .map(depot => {
      const dist = haversineDistance(request.latitude, request.longitude, depot.latitude, depot.longitude);
      return {
        depot_id: depot.item_id,
        depot_name: `${depot.source} (${depot.location})`,
        depot_type: depot.source,
        distance_km: dist,
        available_quantity: depot.quantity,
        can_fulfill_fully: depot.quantity >= request.quantity,
        contact_person: depot.contact_person,
        contact_phone: depot.contact_phone
      };
    });

  // Sort by fulfillment capability and closest distance
  candidates.sort((a, b) => {
    if (a.can_fulfill_fully && !b.can_fulfill_fully) return -1;
    if (!a.can_fulfill_fully && b.can_fulfill_fully) return 1;
    return a.distance_km - b.distance_km;
  });

  if (candidates.length === 0) {
    return {
      match_found: false,
      requested_item: request.required_item,
      required_quantity: request.quantity,
      allocations: [],
      total_allocated: 0,
      remaining_unmet_demand: request.quantity
    };
  }

  const primary = candidates[0];
  const allocations: MatchResult['allocations'] = [];
  let remaining = request.quantity;

  for (const c of candidates) {
    if (remaining <= 0) break;
    const take = Math.min(c.available_quantity, remaining);
    allocations.push({
      depot_id: c.depot_id,
      depot_name: c.depot_name,
      depot_type: c.depot_type,
      distance_km: c.distance_km,
      allocated_quantity: take,
      contact_person: c.contact_person,
      contact_phone: c.contact_phone
    });
    remaining -= take;
  }

  return {
    match_found: true,
    requested_item: request.required_item,
    required_quantity: request.quantity,
    primary_recommendation: {
      depot_name: primary.depot_name,
      depot_type: primary.depot_type,
      distance_km: primary.distance_km,
      available_quantity: primary.available_quantity,
      can_fulfill_fully: primary.can_fulfill_fully
    },
    allocations,
    total_allocated: request.quantity - remaining,
    remaining_unmet_demand: Math.max(0, remaining)
  };
}

export function computeEssentialCoverageScore(
  stateName: string,
  pillarsInput: Record<string, { required: number; available: number }>
): RegionalCoverageScore {
  const pillarsList = ['Food', 'Water', 'Medicine', 'Shelter', 'Sanitation'];
  const pillars: Record<string, { required: number; available: number; shortage: number; coverage_pct: number }> = {};
  
  let totalPct = 0;
  let lowestPillar = 'Food';
  let lowestCoverage = 101;

  for (const p of pillarsList) {
    const data = pillarsInput[p] || { required: 1000, available: 800 };
    const shortage = Math.max(0, data.required - data.available);
    const coverage_pct = Math.min(100, Math.round((data.available / Math.max(1, data.required)) * 100));

    pillars[p] = {
      required: data.required,
      available: data.available,
      shortage,
      coverage_pct
    };

    totalPct += coverage_pct;
    if (coverage_pct < lowestCoverage) {
      lowestCoverage = coverage_pct;
      lowestPillar = p;
    }
  }

  const overall = Math.round(totalPct / pillarsList.length);

  return {
    state: stateName,
    overall_coverage_pct: overall,
    lowest_coverage_pillar: lowestPillar,
    biggest_gap_message: `Critical relief bottleneck identified in ${lowestPillar} (${lowestCoverage}% coverage). Priority dispatch required.`,
    pillars
  };
}

export function runDemandForecast(
  disaster: string,
  severity: 'LOW' | 'MODERATE' | 'HIGH' | 'CRITICAL',
  population: number,
  durationDays: number,
  currentStocks: { food: number; water: number; medicine: number; blankets: number }
) {
  const mult = {
    'CRITICAL': 1.4,
    'HIGH': 1.2,
    'MODERATE': 1.0,
    'LOW': 0.8
  }[severity];

  const reqFood = Math.round(population * 3 * durationDays * mult);
  const reqWater = Math.round(population * 3.5 * durationDays * mult);
  const reqMeds = Math.round(population * 0.05 * mult);
  const reqBlankets = Math.round(population * 0.4 * mult);

  const forecast = [
    { name: 'Food Packets', required: reqFood, current: currentStocks.food, unit: 'packets' },
    { name: 'Drinking Water', required: reqWater, current: currentStocks.water, unit: 'liters' },
    { name: 'Medical Kits', required: reqMeds, current: currentStocks.medicine, unit: 'kits' },
    { name: 'Blankets', required: reqBlankets, current: currentStocks.blankets, unit: 'pieces' }
  ].map(item => {
    const shortage = Math.max(0, item.required - item.current);
    const coverage = Math.min(100, Math.round((item.current / Math.max(1, item.required)) * 100));
    return {
      ...item,
      shortage,
      coverage,
      isCritical: coverage < 50
    };
  });

  const alerts = forecast.filter(f => f.isCritical).map(f => 
    `Critical ${f.name} Shortage: Depletion within ${durationDays * 24}h. Gap: ${f.shortage.toLocaleString()} ${f.unit}`
  );

  return {
    disaster,
    severity,
    population,
    durationDays,
    forecast,
    alerts,
    hasCriticalWarning: alerts.length > 0
  };
}

File: ai/relief-demand/relief_engine.py
"""
AapdaSathi - Member 3: Relief Intelligence & Relief Matching AI Engine
Provides:
1. Relief Priority Scoring (SRS Section 18)
2. Need-to-Relief Distance & Inventory Matching (SRS Section 12)
3. AI-Based Predictive Relief Demand Forecasting (SRS Section 36)
4. "No One Left Without Essentials" Essential Coverage Score (SRS Section 17)
5. Shortage Heatmap Data Generator (SRS Section 13)
"""

import math
from typing import List, Dict, Any, Optional

EARTH_RADIUS_KM = 6371.0

# Predefined coordinates for major North Eastern Region (NER) centres
NER_LOCATIONS = {
    # Assam
    "Guwahati": (26.1445, 91.7362),
    "Dibrugarh": (27.4728, 94.9120),
    "Silchar": (24.8333, 92.7789),
    "Jorhat": (26.7509, 94.2037),
    "Nagaon": (26.3468, 92.6840),
    # Arunachal Pradesh
    "Itanagar": (27.0844, 93.6053),
    "Tawang": (27.5861, 91.8594),
    "Pasighat": (28.0664, 95.3265),
    "Bomdila": (27.2645, 92.4159),
    # Manipur
    "Imphal": (24.8170, 93.9368),
    "Churachandpur": (24.3333, 93.6833),
    "Thoubal": (24.6382, 94.0142),
    # Meghalaya
    "Shillong": (25.5788, 91.8933),
    "Tura": (25.5138, 90.2202),
    "Jowai": (25.4500, 92.2000),
    "Sohra": (25.2986, 91.7324),
    # Mizoram
    "Aizawl": (23.7271, 92.7176),
    "Lunglei": (22.8872, 92.7388),
    "Champhai": (23.4735, 93.3278),
    # Nagaland
    "Kohima": (25.6751, 94.1086),
    "Dimapur": (25.9090, 93.7266),
    "Mokokchung": (26.3253, 94.5204),
    # Tripura
    "Agartala": (23.8315, 91.2868),
    "Dharmanagar": (24.3756, 92.1648),
    "Udaipur": (23.5333, 91.4833),
    # Sikkim
    "Gangtok": (27.3389, 88.6065),
    "Namchi": (27.1667, 88.3500),
    "Mangan": (27.5111, 88.5283)
}

def haversine_distance(lat1: float, lon1: float, lat2: float, lon2: float) -> float:
    """Calculate the great circle distance between two points in km."""
    dlat = math.radians(lat2 - lat1)
    dlon = math.radians(lon2 - lon1)
    a = (math.sin(dlat / 2) ** 2 +
         math.cos(math.radians(lat1)) * math.cos(math.radians(lat2)) *
         math.sin(dlon / 2) ** 2)
    c = 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))
    return round(EARTH_RADIUS_KM * c, 2)


# ==========================================
# 1. RELIEF PRIORITY SCORE (SRS Section 18)
# ==========================================
def calculate_priority_score(
    population: int,
    urgency: str,
    vulnerable_count: Dict[str, int],
    road_blocked: bool = False,
    current_shortage_ratio: float = 0.5
) -> Dict[str, Any]:
    """
    Priority Score = Population Factor + Urgency + Resource Shortage + Accessibility + Vulnerability
    Output normalized score between 0 and 100 with category classification.
    """
    pop_score = min(25.0, (math.log10(max(population, 1)) / 4.0) * 25.0)

    urgency_map = {
        "CRITICAL": 25.0,
        "HIGH": 18.0,
        "MODERATE": 10.0,
        "LOW": 4.0
    }
    urgency_score = urgency_map.get(urgency.upper(), 10.0)
    shortage_score = min(20.0, current_shortage_ratio * 20.0)

    children = vulnerable_count.get("children", 0)
    elderly = vulnerable_count.get("elderly", 0)
    pregnant = vulnerable_count.get("pregnant", 0)
    injured = vulnerable_count.get("injured", 0)
    total_vulnerable = children + elderly + pregnant + injured
    
    if population > 0 and total_vulnerable > 0:
        vuln_density = (children * 1.5 + elderly * 1.5 + pregnant * 2.5 + injured * 3.0) / population
        vuln_score = min(20.0, vuln_density * 40.0)
    else:
        vuln_score = 0.0

    accessibility_score = 10.0 if road_blocked else 2.0

    total_score = round(pop_score + urgency_score + shortage_score + vuln_score + accessibility_score, 1)
    total_score = min(100.0, max(0.0, total_score))

    if total_score >= 80.0:
        priority_class = "CRITICAL"
        badge_color = "red"
    elif total_score >= 60.0:
        priority_class = "HIGH"
        badge_color = "orange"
    elif total_score >= 35.0:
        priority_class = "MODERATE"
        badge_color = "yellow"
    else:
        priority_class = "LOW"
        badge_color = "green"

    return {
        "priority_score": total_score,
        "priority_class": priority_class,
        "badge_color": badge_color,
        "breakdown": {
            "population_score": round(pop_score, 1),
            "urgency_score": round(urgency_score, 1),
            "shortage_score": round(shortage_score, 1),
            "vulnerability_score": round(vuln_score, 1),
            "accessibility_score": round(accessibility_score, 1)
        }
    }


# ==========================================
# 2. NEED-TO-RELIEF MATCHING (SRS Section 12)
# ==========================================
def match_need_to_relief(
    requested_item: str,
    required_quantity: float,
    request_lat: float,
    request_lon: float,
    inventory_depots: List[Dict[str, Any]]
) -> Dict[str, Any]:
    """
    Searches available sources:
    - Government Warehouses
    - Relief Centres
    - NGOs
    - Volunteers & Donors
    - Sibling Relief Camps
    Matches nearest available sources by Haversine distance and stock availability.
    """
    candidates = []

    for depot in inventory_depots:
        items = depot.get("items", {})
        avail_stock = items.get(requested_item, 0)
        
        if avail_stock <= 0:
            continue

        depot_lat = depot["latitude"]
        depot_lon = depot["longitude"]
        dist_km = haversine_distance(request_lat, request_lon, depot_lat, depot_lon)

        candidates.append({
            "depot_id": depot["id"],
            "depot_name": depot["name"],
            "depot_type": depot.get("type", "Warehouse"),
            "state": depot.get("state", "Assam"),
            "district": depot.get("district", "Kamrup"),
            "distance_km": dist_km,
            "available_quantity": avail_stock,
            "can_fulfill_fully": avail_stock >= required_quantity,
            "latitude": depot_lat,
            "longitude": depot_lon,
            "contact_person": depot.get("contact_person", "Officer in charge"),
            "contact_phone": depot.get("contact_phone", "+91 98765 43210")
        })

    candidates.sort(key=lambda x: (not x["can_fulfill_fully"], x["distance_km"]))

    if not candidates:
        return {
            "match_found": False,
            "message": f"No available stock for '{requested_item}' within registered depots.",
            "allocations": [],
            "total_allocated": 0,
            "remaining_unmet_demand": required_quantity
        }

    primary_match = candidates[0]
    allocations = []
    remaining = required_quantity

    for cand in candidates:
        if remaining <= 0:
            break
        qty_to_take = min(cand["available_quantity"], remaining)
        allocations.append({
            "depot_id": cand["depot_id"],
            "depot_name": cand["depot_name"],
            "depot_type": cand["depot_type"],
            "distance_km": cand["distance_km"],
            "allocated_quantity": qty_to_take,
            "contact_person": cand["contact_person"],
            "contact_phone": cand["contact_phone"]
        })
        remaining -= qty_to_take

    return {
        "match_found": True,
        "requested_item": requested_item,
        "required_quantity": required_quantity,
        "primary_recommendation": {
            "depot_name": primary_match["depot_name"],
            "depot_type": primary_match["depot_type"],
            "distance_km": primary_match["distance_km"],
            "available_quantity": primary_match["available_quantity"],
            "can_fulfill_fully": primary_match["can_fulfill_fully"]
        },
        "allocations": allocations,
        "total_allocated": required_quantity - remaining,
        "remaining_unmet_demand": max(0, remaining),
        "candidates_evaluated": len(candidates)
    }


# ========================================================
# 3. PREDICTIVE RELIEF DEMAND FORECASTING (SRS Section 36)
# ========================================================
def forecast_relief_demand(
    disaster_type: str,
    severity: str,
    population_affected: int,
    expected_duration_days: int,
    current_stocks: Dict[str, float]
) -> Dict[str, Any]:
    """
    AI-Based Relief Demand Forecasting:
    Based on population + disaster severity + duration
    """
    sev_mult = {
        "CRITICAL": 1.4,
        "HIGH": 1.2,
        "MODERATE": 1.0,
        "LOW": 0.8
    }.get(severity.upper(), 1.0)

    food_needed = round(population_affected * 3.0 * expected_duration_days * sev_mult)
    water_liters_needed = round(population_affected * 3.5 * expected_duration_days * sev_mult)
    medical_kits_needed = round(population_affected * 0.05 * sev_mult)
    blankets_needed = round(population_affected * 0.4 * sev_mult)

    forecast_items = [
        {"item": "Food Packets", "required": food_needed, "current": current_stocks.get("Food Packets", 0.0), "unit": "packets"},
        {"item": "Drinking Water", "required": water_liters_needed, "current": current_stocks.get("Drinking Water", 0.0), "unit": "liters"},
        {"item": "Medical Kits", "required": medical_kits_needed, "current": current_stocks.get("Medical Kits", 0.0), "unit": "kits"},
        {"item": "Blankets", "required": blankets_needed, "current": current_stocks.get("Blankets", 0.0), "unit": "pieces"}
    ]

    has_critical_shortage = False
    critical_alerts = []

    for f in forecast_items:
        shortage = max(0.0, f["required"] - f["current"])
        f["shortage"] = shortage
        f["coverage_ratio"] = round(min(1.0, f["current"] / max(1.0, f["required"])) * 100, 1)
        if f["coverage_ratio"] < 50.0:
            has_critical_shortage = True
            critical_alerts.append(
                f"Critical {f['item']} shortage: {int(shortage):,} {f['unit']} required within {expected_duration_days * 24} hours."
            )

    return {
        "disaster_type": disaster_type,
        "severity": severity,
        "population_affected": population_affected,
        "duration_days": expected_duration_days,
        "has_critical_shortage": has_critical_shortage,
        "critical_alerts": critical_alerts,
        "forecast_breakdown": forecast_items,
        "summary": (
            f"AI Forecast: {population_affected:,} citizens affected by {severity} {disaster_type} "
            f"will require {food_needed:,} food packets and {water_liters_needed:,}L water over {expected_duration_days} days."
        )
    }


# ========================================================
# 4. "NO ONE LEFT WITHOUT ESSENTIALS" (SRS Section 17)
# ========================================================
def calculate_essential_coverage(
    requirements: Dict[str, Dict[str, float]]
) -> Dict[str, Any]:
    """
    Calculates regional coverage across 5 pillars:
    Food, Water, Medicine, Shelter, Sanitation.
    """
    pillars = ["Food", "Water", "Medicine", "Shelter", "Sanitation"]
    breakdown = {}
    total_coverage = 0.0

    lowest_pillar = None
    lowest_coverage = 101.0

    for pillar in pillars:
        data = requirements.get(pillar, {"required": 1000.0, "available": 1000.0})
        req = max(1.0, data.get("required", 1.0))
        avail = data.get("available", 0.0)
        coverage_pct = round(min(100.0, (avail / req) * 100.0), 1)
        
        breakdown[pillar] = {
            "required": req,
            "available": avail,
            "shortage": max(0.0, req - avail),
            "coverage_pct": coverage_pct
        }
        total_coverage += coverage_pct

        if coverage_pct < lowest_coverage:
            lowest_coverage = coverage_pct
            lowest_pillar = pillar

    overall_coverage = round(total_coverage / len(pillars), 1)

    return {
        "overall_coverage_pct": overall_coverage,
        "lowest_coverage_pillar": lowest_pillar,
        "biggest_gap_message": f"Critical gap detected in {lowest_pillar} at only {lowest_coverage}% fulfillment.",
        "breakdown": breakdown
    }
